optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/QuadTuple.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.common;
```

Comment:
Maybe all Quad variants cloud be in a quad package?
Bavet's impl and the API does this: there is common, uni, bi, tri (and in the future quad).

Reply:
If you open the code in the IDE, you'll see that some of these tuples are being used from more than just one package - therefore I put them all in `common` for the sake of consistency.

Reply:
> I wonder where the natural conclusion of these PRs (bi, tri, quad, pent..) will be!??

I guess either a future Java release supports dynamic number of generic parameters or we reach [the largest natural number](https://simple.wikipedia.org/wiki/Natural_number). Whatever comes first ;-)

The pain we go through so our users can have a strongly typed API... :)